feature documentation & streaming guide updates (#71)
smartdevicelink/sdl_core_guides: docs/Feature Documentation/Audio and Video Streaming Setup/index.md

## Initial Configuration

### SDL Core Setup

> If you don't already have Core built, you can follow the [install and run guide](../../getting-started/install-and-run/).

Reviewer: The install and run guide is only applicable to setting up sdl_core. If there is a guide for setting up the sdl_hmi we should link to that as well. If not, the `Clone the SDL HMI Repository` section should not be removed.

Author: added in https://github.com/smartdevicelink/sdl_core_guides/pull/71/commits/b07a1990e1799725a09996753b1db4bba73b664b

## Initial Configuration

### SDL Core Setup

> If you don't already have Core built, you can follow the [install and run guide](../../getting-started/install-and-run/).

Reviewer 2: ```suggestion
If you don't already have Core built, you can follow the [Install and Run Guide](../../getting-started/install-and-run/).
```

gst-launch-1.0 filesrc location=/path/to/h264/file ! decodebin ! videoconvert ! ximagesink sync=false
```

If you're using tcp, you can connect the stream directly with your phone's IP address using
```
gst-launch-1.0 tcpclientsrc host=<Device IP Address> port=3000 ! decodebin ! videoconvert ! ximagesink sync=false
```

### HMI Modifications
Using either pipe or socket streaming may require a modification to the SDL HMI.

#### VIDEO
Comment out the following line in `app/model/sdl/Abstract/Model.js`:
```
//  SDL.SDLModel.playVideo(appID);
```

#### AUDIO
Comment out the following lines in `app/model/sdl/Abstract/Model.js`:
```
//  SDL.StreamAudio.play(
//      SDL.SDLController.getApplicationModel(appID).navigationAudioStream
//  );
```

## Pipe Streaming

### Configuration (smartDeviceLink.ini)

> In the Core build folder, open bin/smartDeviceLink.ini and the make the following changes:

Reviewer: ```suggestion
In the Core build folder, open `bin/smartDeviceLink.ini` and the make the following changes:
```

Reviewer: Also, the Configuration section for socket streaming should have the same description for modifying the ini file, just to keep it consistent.

Author: good idea, fixed in https://github.com/smartdevicelink/sdl_core_guides/pull/71/commits/b07a1990e1799725a09996753b1db4bba73b664b

AudioStreamConsumer = pipe
```

> #### VIDEO

Reviewer: This subheading should be removed since video socket streaming doesn't have a similar heading and the `####` subheadings are already used for the video/audio format

Author: This header applies to both socket and pipe streaming as the description under it says, I think it makes a lot of sense to preface the VIDEO and AUDIO sections with a header that includes them both, called something like HMI modifications.

I moved the whole section under SDL HMI setup which should help the confusion that it just applies to pipe streaming. Let me know what you think.

Reviewer: Sorry should have been more specific I was refering to the `#### VIDEO` heading

Currently the pipe and socket Streaming sections are setup as 
```md
## <Pipe/Socket> Streaming

### Configuration (smartDeviceLink.ini)
...
### Video Stream <Pipe/Socket>
...
### Audio Stream <Pipe/Socket>
...
```
There is however an extra subheading above the `### Video Stream Pipe` section which is somewhat out of place
![Screen Shot 2020-03-04 at 3 14 51 PM](https://user-images.githubusercontent.com/10703941/75919565-c92d1100-5e2b-11ea-9e0b-1b97f4d83ffd.png). I was saying that we should remove https://github.com/smartdevicelink/sdl_core_guides/pull/71/commits/b07a1990e1799725a09996753b1db4bba73b664b#diff-e3759fc012a839fcb83e8cf8b6408a10R65

Author: AH! great catch. fixed in https://github.com/smartdevicelink/sdl_core_guides/pull/71/commits/f7e530b4c9121bf3adb5c693975b2aa2d0962751

gst-launch-1.0 souphttpsrc location=http://127.0.0.1:5080 ! audio/x-raw,format=S16LE,rate=16000,channels=1 ! pulsesink
```

## Video Streaming States

This section describes how Core manages the streaming states of mobile applications. Only one application may stream video at a time, but audio applications may stream while in the LIMITED state with other applications.

When an app is moved to HMI level `FULL`:

> - All non-streaming applications go to HMI level `BACKGROUND`

Reviewer: The bullet points in https://staging.smartdevicelink.com/en/guides/pull_request/core_6_1/0be31839052bea6e9c529af051d82e2d35201ebc/integrating-your-hmi/sdl-core-and-hmi-communication/#request look somewhat better. I'd suggest trying that format for the bullet points.

```suggestion
  * All non-streaming applications go to HMI level `BACKGROUND`
```

Author: definitely. changed in https://github.com/smartdevicelink/sdl_core_guides/pull/71/commits/b07a1990e1799725a09996753b1db4bba73b664b

Reviewer: Thanks for making the change. Unfortunately the result isn't quite what I was expecting.
![Screen Shot 2020-03-04 at 3 00 54 PM](https://user-images.githubusercontent.com/10703941/75918141-2a9fb080-5e29-11ea-8f6e-939b4fc08cd4.png)

Not sure why the doc server parses it like that. I did notice that there is a new line before the first bullet in https://smartdevicelink.com/en/guides/pull_request/core/c3c6b5e47bc41ba52003b393711775eae3a40ab0/integrating-your-hmi/sdl-core-and-hmi-communication/#request.  If you want to experiment with it, you could try making that change. Otherwise I think it's okay to change it back to `-`

- All apps with the same App HMI Type go to `BACKGROUND`
- Streaming apps with a different App HMI Type that were in `FULL` go to `LIMITED`

When an app is moved to HMI level `LIMITED`:
- All non-streaming applications keep their HMI level
- All applications with a different App HMI Type keep their HMI level
- Applications with the same App HMI Type go to `BACKGROUND`

### Start the Web HMI

CD into the HMI repository and run
```
chromium-browser index.html
```

### Additional Resources

!!! NOTE

> Livio provides an [example video streaming android application](https://github.com/livio/sdl_video_streaming_android_sample).

Reviewer: ```suggestion
Livio provides a [sample video streaming android application](https://github.com/livio/sdl_video_streaming_android_sample). This app is for testing purposes only.
```

Author: sample is in the name of the app so i like example more, and testing purposes only is in the readme, why do you think we should add here too? i think that might unnecessarily scare people off from using it as a reference.

Reviewer: I was thinking it would make it clearer but yeah I guess `example` should make it apparent that the app is only meant for testing. The change might be uneccesary

!!!

> [iOS Video Streaming Guide](../../../iOS/video-streaming-for-navigation-apps/video-streaming/)

Reviewer: This link doesn't work when viewing it in a PR. I think it should just be a static link 
```suggestion
[iOS Video Streaming Guide](https://smartdevicelink.com/en/guides/iOS/video-streaming-for-navigation-apps/video-streaming/)
```

Author: on PR it adds two more levels (/pull_request and /$hash) so I tested the links adding two more `../` to the relative paths and they worked fine.

Reviewer: I'm viewing it in with the `6.1 Release Overhaul Testing` version. I think the link should be fixed so it can work consistently, regardless of whether it's in a PR or in production

Author: But a full path will not work properly in staging... it's a trade off either way but I think a relative path is the correct thing to do here, so any of the server config can change and it can still work or so we can host a development version (staging) of the guides. I really think using a full path in a situation like this is just the lazy option. If you're sure the correct trade off is to use a full path let me know and I'll change it.

Reviewer: hmm though decision but I think for now for the sake of consistency and allowing it to work in PRs I think the full link might be better for now

Author: added in https://github.com/smartdevicelink/sdl_core_guides/pull/71/commits/9f48ec382fbbe03a31807656c63517457e3cf268

> [Android Video Streaming Guide](../../../android/video-streaming-for-navigation-apps/video-streaming/)

Reviewer: ```suggestion
[Android Video Streaming Guide](https://smartdevicelink.com/en/guides/android/video-streaming-for-navigation-apps/video-streaming/)
```

Author: The relative links should work once this is merged, any reason that full paths are better?

Reviewer: Full paths would be better in this case since this is basically an external resource(not part of sdl_core_guides) and the link would always work regardless of any modifications to our local config structure

Also, we usually only use relative links for local resources. For linking to external guides we use the full path( for example the link to the policy overview https://github.com/smartdevicelink/sdl_core_guides/pull/65/files#diff-4d81113b72a417a09e4e6b1cb2028e32R50)

Author: The iOS guide is a local resource, they're both on the same hard drive hosted by the same server on the same domain. Using a full path will break the link on sites like staging. If you're sure the correct trade off is to use a full path let me know and I'll change it.

Reviewer: https://github.com/smartdevicelink/sdl_core_guides/pull/71#discussion_r388485301

Author: added in https://github.com/smartdevicelink/sdl_core_guides/pull/71/commits/9f48ec382fbbe03a31807656c63517457e3cf268
